Actions and Events for interpolating without interp component

markscheel opened the pull request:
## Proposed changes

Adds a new Action and a new Event that will interpolate without the use of an Interpolator
component.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Add new InterpolateToTarget Action.

markscheel committed:
Add Test_InterpolateToTarget

markscheel committed:
Add InterpolateWithoutInterpComponent.hpp and test.

nilsdeppe commented:
@markscheel this is failing to build in CI so I'm marking as in progress for now

markscheel commented:
@nilsdeppe agreed about in-progress.  It builds fine with clang, but not with gcc.  I have a fix that makes it build with gcc but one of the tests segfaults with gcc. (Again clang is ok).  It seems to be unhappy with making temporary databoxes.

nilsdeppe commented:
Ah okay. So I know @kidder made some changes a month or so back that I think eliminated aliasing DataBoxes or something like that. If it would be helpful we can have a hangout to discuss. Just let me know! :)

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp:
```diff
@@ -0,0 +1,116 @@
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Utilities/TMPL.hpp"
+
```
forward declaration for TaggedTuple

kidder replied:
just need a forward declaration as it is only used as a commented out function parameter reference (which you could also make a const reference)

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp:
```diff
@@ -0,0 +1,116 @@
+
+    // 1. Create a new DataBox that contains
+    // InterpolationTargetTag::compute_items_on_source
+    auto new_box = db::create_from<
```
make sure this works properly if any of the tags in `db::AddComputeTags<typename InterpolationTargetTag::compute_items_on_source>>` are all ready in`box`

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp:
```diff
@@ -0,0 +1,116 @@
+
+    tmpl::for_each<
+        typename InterpolationTargetTag::vars_to_interpolate_to_target>(
+        [&new_box, &local_vars ](auto x) noexcept {
```
x -> `tag_v`

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp:
```diff
@@ -0,0 +1,116 @@
+    Parallel::simple_action<
+        Actions::InterpolationTargetVarsFromElement<InterpolationTargetTag>>(
+        receiver_proxy,
+        std::vector<Variables<
```
would it be better to just interpolate each tensor of the variables in the `tmpl::for_each` above instead of copying twice?

markscheel replied:
What do you mean by copying twice? I see one "obvious" copy, the one in the for_each that copies each tensor into a Variables.  Is the other copy you are referring to the call to the simple_action, which copies its arguments?  Or do you mean that the std::vector in line 100 is being filled by copying the result of interpolator.interpolate instead of moving that result (in which case I should construct the vector differently on line 100)?

Are you suggesting to add a new member function to interp::Irregular that takes a Tensor instead of a Variables, so that the copy in the for_each can be eliminated?

nilsdeppe replied:
Sorry, there was some miscommunication between Larry and I. There is only one additional copy being done. The question is: if the interpolant would support interpolating a `Tensor` by `not_null`, would it be more efficient to just do the interpolation directly rather than copying into a buffer, then interpolating. There are currently 2 memory allocations, which could be reduced to one, but that's unlikely to be an issue.

markscheel replied:
Maybe.  If the interpolant could interpolate a Tensor by not_null, then there would be one
fewer memory allocation.  But then the interpolator would need to be called for each Tensor individually, instead of doing the entire Variables as a single matrix operation.  That is, it would be several matrix-vector multiplies instead of one matrix-matrix multiply.

nilsdeppe replied:
Discussion in SpECTRE call: filing an issue and deferring is fine

nilsdeppe replied:
Did you file an issue for this already @markscheel?

markscheel replied:
Yes, #2271

kidder commented on tests/Unit/Helpers/NumericalAlgorithms/Interpolation/InterpolateOnElementTestHelpers.hpp:
```diff
@@ -0,0 +1,251 @@
+    // For this tag, supply twice the analytic solution
+    get(solution) =
+        2.0 *
+        (2.0 * get<0>(coords) + 3.0 * get<1>(coords) + 5.0 * get<2>(coords));
```
just multiply the coefficients by two

kidder commented on tests/Unit/Helpers/NumericalAlgorithms/Interpolation/InterpolateOnElementTestHelpers.hpp:
```diff
@@ -0,0 +1,251 @@
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Initialization,
+          tmpl::list<ActionTesting::InitializeDataBox<simple_tags>>>,
+      Parallel::PhaseActions<typename Metavariables::Phase,
```
don't need to have phases with empty lists (as long as one phase has a list)

kidder commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolateToTarget.cpp:
```diff
@@ -0,0 +1,129 @@
+  enum class Phase { Initialization, Testing, Exit };
+};
+
+struct MockMetavariablesWithoutComputeItemOnSource {
```
I don't understand the difference between these two structs...

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,189 @@
+}  // namespace intrp
+/// \endcond
+
+namespace intrp {
```
/// cond all the forward declarations

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,189 @@
+
+/// Does an interpolation onto an InterpolationTargetTag by calling Actions on
+/// the InterpolationTarget component.
+template <size_t VolumeDim, typename InterpolationTargetTag,
```
just move this forward declaration above to replace the other one

markscheel replied:
Sorry, I don't understand what you mean.  Does "the other one" mean the forward declaration on line 48?  I don't understand how that can be replaced because Registrars::InterpolateWithoutInterpComponent depends on it (and the forward declaration on line 65 depends on Registrars::InterpolateWithoutInterpComponent).

kidder replied:
I'm surprised this works; I would forward declare `Registrars::InterpolateWithoutInterpComponent`, then this forward declaration, and then the definitions...

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,189 @@
+
+  using options = tmpl::list<>;
+  static constexpr OptionString help =
+      "Starts interpolation onto the given InterpolationTargetTag.";
```
what do you mean by "Starts" ?  Note that help strings no longer have a character limit, and they are auto-formatted...

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,189 @@
+                                 compute_items_on_source>) {
+      // 1.a Copy the tensors directly into the variables; no need to
+      // make a DataBox because we have no ComputeItems.
+      const auto copy_to_variables = [&interp_vars](
```
try using attribute `[[maybe-unused]]`

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,189 @@
+      // Copy vars_to_interpolate_to_target from databox to vars
+      tmpl::for_each<
+          typename InterpolationTargetTag::vars_to_interpolate_to_target>(
+          [&box, &interp_vars ](auto x) noexcept {
```
x -> `tag_v`

kidder replied:
ping

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,189 @@
+    Parallel::simple_action<
+        Actions::InterpolationTargetVarsFromElement<InterpolationTargetTag>>(
+        receiver_proxy,
+        std::vector<Variables<
```
same question about interpolating in the `tmpl::for_each` to avoid copies

kidder commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolateWithoutInterpComponent.cpp:
```diff
@@ -0,0 +1,136 @@
+  enum class Phase { Initialization, Testing, Exit };
+};
+
+struct MockMetavariablesWithoutComputeItemsOnSource {
```
combine these with a bool template argument

kidder reviewed (changes requested):
reviewed with @nilsdeppe

markscheel committed:
Test_TMPL: Add test for list_difference with empty 1st arg.

markscheel committed:
Make clang-tidy happy.

markscheel commented:
Made fixups, and rebased onto develop.

There is one clang-tidy error that @nilsdeppe  has fixed elsewhere but will still fail here.

kidder commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -41,14 +41,23 @@ class ElementId;
-          typename Metavariables, typename Tensors, typename EventRegistrars>
-class InterpolateWithoutInterpComponent;
 
 namespace Registrars {
```
you should document or cond out the Registrars

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp:
```diff
@@ -68,13 +68,15 @@ struct InterpolateToTarget {
     // in several steps:
     const auto& element_coord_holder = element_coord_holders.at(element_ids[0]);
 
+    using compute_items_in_new_box = tmpl::list_difference<
```
is this tested with a an item already in the box?

markscheel replied:
I'm adding such a test now...

nilsdeppe replied:
Could you point us to the place where it's tested? That'll make it easier to review :)

nilsdeppe commented:
The clang-tidy issue will be fixed if you rebase on develop

kidder reviewed (changes requested):
please rebase and squash including our these comments.  Your clang-tidy error should go away with the rebase

markscheel commented:
Made changes, rebased, squashed.

nilsdeppe commented on tests/Unit/Helpers/NumericalAlgorithms/Interpolation/InterpolateOnElementTestHelpers.hpp:
```diff
@@ -0,0 +1,243 @@
+  get(solution) =
+      (2.0 * get<0>(coords) + 3.0 * get<1>(coords) + 5.0 * get<2>(coords));
+
+  // NOLINTNEXTLINE clang-tidy is unhappy with `if constexpr` braces.
```
Is this still true with clang-tidy 10?

markscheel replied:
Cool. clang-tidy 10 now works. Removing the NOLINTS...

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/Events/InterpolateWithoutInterpComponent.hpp:
```diff
@@ -0,0 +1,193 @@
+    Variables<typename InterpolationTargetTag::vars_to_interpolate_to_target>
+        interp_vars(mesh.number_of_grid_points());
+
+    // Clang-tidy wants extra braces for `if constexpr`
```
Is this still true with clang-tidy 10?

markscheel replied:
here too, clang-tidy works again.  Removing NOLINTs,,,

markscheel commented:
The newer test is in Test_interpolateToTarget.  In there, mock_element has an additional template parameter "AddComputeItemToBox", and MockMetavariables has the same template parameter. Then you will see at for the test, all combinations of 'use the computeitem' and 'add it to the box beforehand' are tested.

Removed the NOLINTS and squashed.
